### PR TITLE
[GEP-19] Migrate shoot Alertmanager deployment and configuration

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area dev-productivity monitoring
/kind enhancement

**What this PR does / why we need it**:
Similar to https://github.com/gardener/gardener/pull/9159, this PR migrates management of the shoot Alertmanager deployment and its configuration to `prometheus-operator`.

The consideration of the new `ManagedResource` in the `shoot-care` controller will be handled separately (today, we have already some `ManagedResource`s with `.spec.class=seed` in the shoot namespace (e.g., for `vali`/`plutono`/...) that are not considered yet).

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/9065

**Special notes for your reviewer**:
/cc @oliver-goetz @ScheererJ 
FYI @istvanballok @rickardsjp 

**Release note**:

```feature developer
NONE
```

### CODE REVIEW

File: pkg/component/monitoring/alertmanager/alertmanager.go [hunk 1]
```diff
@@ -50,9 +50,6 @@ func (a *alertManager) alertManager(takeOverOldPV bool) *monitoringv1.Alertmanag
 					corev1.ResourceCPU:    resource.MustParse("5m"),
 					corev1.ResourceMemory: resource.MustParse("20Mi"),
 				},
-				Limits: corev1.ResourceList{
-					corev1.ResourceMemory: resource.MustParse("200Mi"),
-				},
```
Comment (ScheererJ) [inline pkg/component/monitoring/alertmanager/alertmanager.go]: With prometheus, the restore during start-up can require temporarily a lot of memory, which can cause the out-of-memory situations and hence require to scale/remove the memory limits. Is this the same for alert-manager? I am not very familiar with it, but I thought that there is no much storage in alert-manager. Hence, the problem would not be that pressing here. My assumptions may be wrong, though.

File: pkg/component/monitoring/alertmanager/component.go [hunk 2]
```diff
@@ -36,6 +37,15 @@ const (
 	PortNameMetrics = "metrics"
 )
 
+// Interface contains functions for an alertmanager deployer.
+type Interface interface {
+	component.DeployWaiter
+	// SetIngressAuthSecret sets the ingress auth secret name.
+	SetIngressAuthSecret(*corev1.Secret)
+	// SetIngressWildcardCertSecret sets the ingress wild card cert secret name.
+	SetIngressWildcardCertSecret(*corev1.Secret)
```
Comment (ScheererJ) [inline pkg/component/monitoring/alertmanager/component.go]: The camel case method name suggests that we see wildcard as a single word even if "wild card" also seems to be valid in the English language.

```suggestion
	// SetIngressWildcardCertSecret sets the ingress wildcard cert secret name.
	SetIngressWildcardCertSecret(*corev1.Secret)
```

File: pkg/component/monitoring/alertmanager/component.go [hunk 3]
```diff
@@ -54,14 +64,30 @@ type Values struct {
 	Replicas int32
 	// AlertingSMTPSecret is the alerting SMTP secret.
 	AlertingSMTPSecret *corev1.Secret
+	// IngressValues contains configuration for exposing this AlertManager instance via an Ingress resource.
+	Ingress *IngressValues
```
Comment (ScheererJ) [inline pkg/component/monitoring/alertmanager/component.go]:
```suggestion
	// Ingress contains configuration for exposing this AlertManager instance via an Ingress resource.
	Ingress *IngressValues
```

File: pkg/component/monitoring/charts/seed-monitoring/charts/alertmanager/templates/alertmanager.yaml [hunk 4]
```diff
@@ -1,175 +0,0 @@
...
-  labels:
-    component: alertmanager
-    role: monitoring
-spec:
-  ports:
-  - port: 9093
-    name: cluster
-  type: ClusterIP
-  # This is important! Without it the mech won't work.
-  clusterIP: None
```
Comment (oliver-goetz) [inline pkg/component/monitoring/charts/seed-monitoring/charts/alertmanager/templates/alertmanager.yaml]: This seems to be important in the past, but I don't find anything similar in the new version.
Does the operator create the service by default the right way?

**Review by oliver-goetz — COMMENTED** [review]
Awesome, this is again a pretty well structured PR 🚀 

I have one question only.

File: pkg/component/monitoring/charts/seed-monitoring/charts/alertmanager/templates/alertmanager.yaml (hunk 4 above)
Comment (rfranzke) [reply pkg/component/monitoring/charts/seed-monitoring/charts/alertmanager/templates/alertmanager.yaml]: Yep, it does :) ref https://github.com/gardener/gardener/pull/9159#discussion_r1492242630

File: pkg/component/monitoring/alertmanager/component.go [hunk 5]
```diff
@@ -54,14 +64,30 @@ type Values struct {
...
+	// AuthSecretName is the name of the auth secret.
+	AuthSecretName string
+	// Host is the hostname under which the AlertManager instance should be exposed.
+	Host string
+	// SecretsManager is the secrets manager used for generating the TLS certificate if no wildcard certificate is
+	// provided.
+	SecretsManager secretsmanager.Interface
+	// WildcardCertName is name of wildcard TLS certificate which is issued for the ingress domain. If not provided, a
+	// self-signed server certificate will be created.
+	WildcardCertName *string
```
Comment (ScheererJ) [inline pkg/component/monitoring/alertmanager/component.go]: This is slightly confusing to me as it seems to be the name of the secret, which contains the certificate, and not the name of the certificate.
WDYT?

**Review by ScheererJ — COMMENTED** [review]
Thanks a lot for migrating shoot alert-manager. I have a few questions.

File: pkg/component/monitoring/charts/seed-monitoring/charts/alertmanager/templates/alertmanager.yaml (hunk 4 above)
Comment (oliver-goetz) [reply pkg/component/monitoring/charts/seed-monitoring/charts/alertmanager/templates/alertmanager.yaml]: There is already an answer for everything, thanks 😄

File: pkg/component/monitoring/alertmanager/alertmanager.go (hunk 1 above)
Comment (rfranzke) [reply pkg/component/monitoring/alertmanager/alertmanager.go]: I don't think the situation is similar, though, I wanted to keep the configuration in sync after https://github.com/gardener/gardener/pull/9205. I don't have a strong opinion. I can also revert this commit if you wish.

Comment (ScheererJ) [reply pkg/component/monitoring/alertmanager/alertmanager.go]: I think it is not that important. As far as I know, `prometheus` can use a lot of memory on startup, but `alert-manager` is not known for this behaviour.
Let's keep it as is and change if required later.

**Review by ScheererJ — COMMENTED** [review]
/lgtm

**Pull request merged**